[subplugins] Refactor subplugin interface

kparichay opened the pull request:
Refactor subplugin interface
Add new interface for the subplugins along with the old interface
Added version number to check if old subplugin interface is used or the new one

Both the interfaces are added as a union,
This will allow for staggered changes rather a big change to be made at once

Related Issue: #2045 

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [*]Skipped
2. Run test: [ ]Passed [ ]Failed [*]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

kparichay commented on ext/nnstreamer/tensor_filter/tensor_filter_cpp.cc:
```diff
@@ -241,6 +237,13 @@ void fini_filter_cpp (void) __attribute__ ((destructor));
 void
 init_filter_cpp (void)
 {
+  NNS_support_cpp.allow_in_place = FALSE;      /** @todo: support this to optimize performance later. */
```
Should this be changed with `extern C` wrapper as C supports non-trivial designated initializers.

myungjoo replied:
No. it's ok with init_filter_cpp. If you are concerned with the readability, you may move init_filter_cpp to Line 55.

kparichay replied:
Moved to line 55 for readability.

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -111,38 +111,106 @@ typedef struct _GstTensorFilterProperties
+ */
+typedef struct _GstTensorFilterFramework
+{
+  uint64_t version; /**< Version of the struct, can be 0x0 (earlier version (_GstTensorFilterFramework_v1)) or 0x2 (newer version (_GstTensorFilterFramework_v2)) */
```
In order to verify the validity in case this struct is given as void * or from dlopen, let's define version as

```
| 32bit (validity check) | 16bit (API version) | 16bit (Subplugin's internal version. Tensor_filter does not case) |
```

Then, let's define:
```
#define GstTensorFilterFrameworkV1 (0xDEAFDEAD00010000)
```
And let the current filters use this macro instead of hardcoded values.

kparichay replied:
Added.

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -111,38 +111,106 @@ typedef struct _GstTensorFilterProperties
+    /**
+     * @brief Tensor_Filter Subplugin definition Version 0
+     */
+    struct /** _GstTensorFilterFramework_v1 */
```
s/v1/v0/

kparichay replied:
done `s/Version 1/Version 0/` and `s/Version 2/Version 1/`

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -184,41 +252,81 @@ typedef struct _GstTensorFilterFramework
+     */
+    struct /** _GstTensorFilterFramework_v2 */
+    {
+      int (*getFrameworkInfo) (void **private_data, GstTensorFilterFrameworkInfo *fw_info);
```
I don't think private_data is available for getFrameworkInfo.
prop will be available at that time, though.

kparichay replied:
private_data is optional (it can be NULL).
this is needed for custom filter subplugin as `allocate_in_invoke` property for custom filter is finalized based on the model. 
OR 
remove private_data from here, and move `allocate_in_invoke` property to getModelInfo (note that this property is model dependent only for custom filter subplugin).

myungjoo replied:
Ah.. ok.. then, let's keep both prop and private_data. and write a "caution" statement as a comment that it could be NULL if it's not yet opened.

kparichay replied:
Added an extra caution statement apart from argument description.

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -184,41 +252,81 @@ typedef struct _GstTensorFilterFramework
+      /**< Mandatory callback. Gets the model related tensor info.
+       * If ops == GET_INPUT_INFO, in_info would contain the input tensor info.
+       * If ops == GET_OUTPUT_INFO, out_info would contain the output tensor info.
+       * If ops == SET_INPUT_INFO, in_info would contain the provided input tensor info, and out_info would contain the updated output tensor info.
```
It appears that you were right about this yesterday. I looks like that we are ok to merge GET-INPUT and GET-OUTPUT as a single op.

kparichay replied:
Merged into a single OP and renamed them,

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -184,41 +252,81 @@ typedef struct _GstTensorFilterFramework
+       * @return 0 if OK. non-zero if error. -ENOENT is operation is not supported.
+       */
+
+      int (*sendEvent) (const GstTensorFilterProperties * prop, void **private_data,
```
Wouldn't ```eventHandler``` be more intuitive for developers?

kparichay replied:
Updated.

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -111,38 +111,106 @@ typedef struct _GstTensorFilterProperties
+   * @return 0 if ok. < 0 if error.
+   */
+
+  void (*close) (const GstTensorFilterProperties * prop, void **private_data);
```
Another thing you need to verify is: Do we really need double pointers for private_data for all callbacks? Don't we need it for open/close only?

kparichay replied:
Yes, I updated this for v1.

kparichay commented:
**V2:**
- Merged operation GET_INPUT_INFO and GET_OUTPUT_INFO into 1. Now atleast one of SET_INPUT_INFO and GET_INOUT_INFO must be supported.
- Added framework versioning helper macros
- Older framework is now version 0 and newer one is version 1
- Renamed sendEvent to eventHandler
- private_data is a double pointer only in open/close in v1
- Invoke in v1 compared to invoke_NN in v0 where private_data is not a double pointer
- Moved init and fini functions in tensor_filter_cpp for readability

jaeyun-jung commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -184,41 +264,93 @@ typedef struct _GstTensorFilterFramework
 
-  int (*checkAvailability) (accl_hw hw);
-      /**< Optional. Check if the provided hardware accelerator is supported. This check is static or dynamic based on framework support. Positive response of this check does not guarantee successful running of model with this accelerator. The static check can be performed without opening the framework.
+      int (*eventHandler) (const GstTensorFilterProperties * prop, void *private_data,
```
Discussion:
Is it necessary to pass any parameter? (e.g. define new event to change accel : CHANGE_ACCEL / GPU)

kparichay replied:
the parameter to be passed via `void *data` is optional (can be NULL).
The handler event will interpret and use `void *data` as per its need and implementation. 
In the example of change accel, `void *data` will have to interpreted as `accl_hw` or `char *`.
As additional operations should directly be able to add into `enum event_ops`, its `eventHandler` is kept generic.

Updated description of data to allow NULL.

myungjoo replied:
void * should be enough for the basic need.

However, we may be able to define a standard struct for each event type as a union of multiple struct per event type. The data structure for each event is not determined by sub-plugins; it's determined by tensor_filter and tensor_filter calls eventHandler.

kparichay replied:
Right, it is determined by the caller and not by the callee.
Added standard struct for each event type.

kparichay commented:
**V4:**
Added to description - eventHandler argument `void *data` is allowed to be NULL

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -111,27 +120,98 @@ typedef struct _GstTensorFilterProperties
-  const char *accl_str; /**< accelerator configuration passed in as parameter */
+  union {
+    const char **supported_accl; /**< accelerators supported by framework intersected with user provided accelerator preference */
+    const char *accl_str; /**< accelerator configuration passed in as parameter */
```
Why these two values are in a union? ```supported_accl``` is a value configured by a subplugin and ```accl_str``` is a value configured by a user (pipeline), right?

kparichay replied:
Yes, these two values are in union because v0 uses `accl_str` (current implementation), and v1 will use `supported_accl`.
In v1, `tensor_filter.c` will call `getFrameworkInfo` to get the list of all the supported frameworks. `tensor_filter.c` will intersect this with user passed list of accelerators, and the intersection (`supported_accl`) will be passed to subplugin when opening it. Subplugin does not directly need to know the exact value configured by the user in the pipeline (`accl_str`). 

Regarding dynamically updated property values, I think they can be best handled with sendEvent (`updateProperty`).

myungjoo replied:
If a given single version will use only one part of the union, it's ok.

kparichay replied:
done.

myungjoo commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -40,6 +40,15 @@
+ */
+#define checkGstTensorFitlerFrameworkVersion(value, version) \
+  ((0xDEAFDEAD00000000ULL | ((version) << 16)) == (value & 0xFFFFFFFFFFFF0000ULL))
+
```
Let's add
```
#define GST_TENSOR_FILTER_API_VERSION_DEFINED (1)
#define GST_TENSOR_FILTER_API_VERSION_MIN (0)
#define GST_TENSOR_FILTER_API_VERSION_MAX (1)
```

So that obsolete external (subplugins implemented in other git repos) code may detect it at build time.

kparichay replied:
Updated. However, currently set `GST_TENSOR_FILTER_API_VERSION_DEFINED` to `0`, and `GST_TENSOR_FILTER_API_VERSION_MAX` to `0`.
Will upgrade these  to 1 once tensor_filter.c supports the new subplugin API.

kparichay commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -111,27 +120,98 @@ typedef struct _GstTensorFilterProperties
+ */
+typedef enum
+{
+  GET_INOUT_INFO,   /**< Gets the input and output tensor info */
```
Agreed, updated.

kparichay commented:
**V5:**
Added tensor filter API versions for external subplugins to verify compatibility at compile time
Renamed GET_INOUT_INFO to GET_IN_OUT_INFO for readability

kparichay commented:
**V6:**
Added more operations to be supported with eventHandler which allow updating the properties
Added GstTensorFilterFrameworkEventData which allows arguments to be passed dependent on the event
GstTensorFilterProperties is passed to all the callbacks

kparichay commented:
**V7:**
Updated destroyNotify to take private_data as an input for V0, and corresponding changes in tensor_filter and subplugins
Added allocateInInvoke for V0
These changes are taken from #2013

myungjoo commented on ext/nnstreamer/tensor_filter/tensor_filter_caffe2.cc:
```diff
@@ -559,32 +559,35 @@ caffe2_getOutputDim (const GstTensorFilterProperties * prop,
-  .destroyNotify = caffe2_destroyNotify,
+  {
+    .name = filter_subplugin_caffe2,
+  }
```
Doesn't this trigger error/warning with SVACE/Coverity?

kparichay replied:
Moved `name` assignment to inside the `ini_filter_*`.

dongju-chae commented on gst/nnstreamer/nnstreamer_plugin_api_filter.h:
```diff
@@ -40,6 +40,20 @@
+ * @brief Check the value of the version field of GstTensorFilterFramework
+ */
+#define checkGstTensorFitlerFrameworkVersion(value, version) \
+  ((0xDEAFDEAD00000000ULL | ((version) << 16)) == (value & 0xFFFFFFFFFFFF0000ULL))
```
How about defining `0xDEAFDEAD00000000UL` for readability? (and can be used in both `GST_TENSOR_FILTER_FRAMEWORK_V0/1`)

myungjoo replied:
Appears good.

kparichay replied:
Updated.

myungjoo commented:
This is a huge change that mandates to change version number (#2096). I'd need more reviews/approvals from reviewers.

Those who have written filter-subplugin: @jaeyun-jung @wooksong @helloahn @leemgs

helloahn commented:
LGTM
